src/app-nav-bar/types.js

```diff
@@ -0,0 +1,31 @@
+/*
+Copyright (c) 2018-2020 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+
+import * as React from 'react';
+
+export type AppNavBarPropsT = {
```

Comment:
Below is a suggestion to refine the types and avoid `any` definitions. I probably missed some nuance, so please let me know where my exploration of the diff went awry.

```js
// eslint-disable-next-line flowtype/no-weak-types
type ItemT = any;

type MainNavItemT = {|
  active?: boolean,
  icon?: React.AbstractComponent<{||}>,
  item: ItemT,
  // maybe include this to customize the menu adapter children?
  mapItemToNode?: React.AbstractComponent<{|item: ItemT|}>,
  mapItemToString: ItemT => string,
  subnav?: MainNavItemT[],
  subnavExitIcon?: React.AbstractComponent<{||}>,
|};

type UserNavItemT = {|
  item: ItemT,
  // maybe include this to customize the menu adapter children?
  mapItemToNode?: React.AbstractComponent<{|item: ItemT|}>,
  mapItemToString: ItemT => string,
|};

export type AppNavBarPropsT = {|
  displayName: string | React.AbstractComponent<{||}>,
  mainNavItems: MainNavItemT[],
  onMainNavItemSelect?: (params: {|item: MainNavItemT|}) => mixed,
  onUserNavItemSelect?: (params: {|item: UserNavItemT|}) => mixed,
  userNavItems?: UserNavItemT[],
  username?: string | React.AbstractComponent<{||}>,
  usernameSubtitle?: string | React.AbstractComponent<{||}>,
  userImgUrl?: string,
|};

export type UserMenuPropsT = {|
  username?: string | React.AbstractComponent<{||}>,
  usernameSubtitle?: string | React.AbstractComponent<{||}>,
  userImgUrl?: string,
|};
```

Reply:
Added the items types per your suggestion, that seem to work pretty well. But kept `React.Node` for those that accept a string or a node as is since the AbstractComponent type seems to be not the right fit for it.